File: apps/api/app/schemas/geo.py
"""Geocoding/행정구역 API schema — `docs/api/regions.md` + `docs/integrations/kor-travel-geo.md`.

kor-travel-geo v2 REST 후보(candidate)는 풍부하고 자주 진화하므로 Pinvi 응답은 candidate를
`dict` 그대로 pass-through하되, 최상위 envelope만 타입 고정한다. 좌표는 `(lon, lat)`.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

SearchKind = Literal["address", "place", "district", "road", "category"]
# kor-travel-geo v2 행정구역 level 어휘 (구 `legal_dong` → `emd`, ADR-049 / geo ADR-056·062).
BoundaryLevel = Literal["sido", "sigungu", "emd"]
RegionRelation = Literal["contains", "overlaps"]


class GeoCandidateList(BaseModel):
    """v2 reverse/geocode/search 공통 — candidate pass-through + 상태."""

    status: str
    candidates: list[dict[str, Any]] = Field(default_factory=list)
    total: int | None = None


class RegionCovering(BaseModel):
    """좌표를 포함하는 행정구역(단건). kor-travel-geo `region` 객체 pass-through."""

    boundary_level: BoundaryLevel
    region: dict[str, Any]


class RegionWithinRadiusItem(BaseModel):
    """반경 내 행정구역 1건. `relation`: contains(중심 포함) | overlaps(반경 원과 교차)."""

    code: str
    name: str | None = None
    relation: RegionRelation


class RegionsWithinRadius(BaseModel):
    """좌표 반경 내 행정구역 — level별 그룹. kor-travel-geo v2 `/v2/regions/within-radius`.

    응답은 후보(candidate) 목록이 아니라 sido/sigungu/emd level별 배열이다(geo ADR-062).
    `center`는 `{lon, lat}`, 요청하지 않은 level은 빈 배열로 온다.
    """

    center: dict[str, Any] = Field(default_factory=dict)
    radius_km: float
    sido: list[RegionWithinRadiusItem] = Field(default_factory=list)
    sigungu: list[RegionWithinRadiusItem] = Field(default_factory=list)
    emd: list[RegionWithinRadiusItem] = Field(default_factory=list)


class UnifiedSearchResult(BaseModel):
    """통합 검색 — feature(kor_travel_map) + address(kor_travel_geo) + 내 POI(Pinvi). 소스별 degrade 가능."""

    features: list[dict[str, Any]] = Field(default_factory=list)
    addresses: list[dict[str, Any]] = Field(default_factory=list)
    my_pois: list[dict[str, Any]] = Field(default_factory=list)
    degraded_sources: list[str] = Field(default_factory=list)

File: packages/schemas/src/geo.ts
import { z } from 'zod';

/** kor-travel-geo v2 candidate는 풍부/가변이라 pass-through(record). 좌표는 (lon, lat). */
const CandidateSchema = z.record(z.string(), z.unknown());

export const BoundaryLevelSchema = z.enum(['sido', 'sigungu', 'legal_dong']);
export type BoundaryLevel = z.infer<typeof BoundaryLevelSchema>;

export const GeoSearchKindSchema = z.enum(['address', 'place', 'district', 'road', 'category']);
export type GeoSearchKind = z.infer<typeof GeoSearchKindSchema>;

/** `/geo/{geocode,reverse,search}` + `/regions/within-radius` 공통 응답. */
export const GeoCandidateListSchema = z.object({
  status: z.string(),
  candidates: z.array(CandidateSchema),
  total: z.number().int().nullable().optional(),
});
export type GeoCandidateList = z.infer<typeof GeoCandidateListSchema>;

/** `/regions/covering-point` 단건 행정구역. */
export const RegionCoveringSchema = z.object({
  boundary_level: BoundaryLevelSchema,
  region: CandidateSchema,
});
export type RegionCovering = z.infer<typeof RegionCoveringSchema>;

/** 통합 검색 결과 소스 태그(ADR-054). internal(feature/my_poi/address) + 외부(kakao/naver). */
export const PlaceSearchSourceSchema = z.enum(['feature', 'my_poi', 'address', 'kakao', 'naver']);
export type PlaceSearchSource = z.infer<typeof PlaceSearchSourceSchema>;

export const PlaceCoordSchema = z.object({ lon: z.number(), lat: z.number() });
export type PlaceCoord = z.infer<typeof PlaceCoordSchema>;

/**
 * `GET /search`의 한 행. source에 따라 채워지는 필드가 다르다. kakao/naver row의 phone/category
 * 등 provider 파생 콘텐츠는 표시 전용이며 저장 대상은 external_ref뿐이다(ADR-054 §7).
 */
export const PlaceSearchResultSchema = z.object({
  source: PlaceSearchSourceSchema,
  name: z.string(),
  coord: PlaceCoordSchema.nullable().default(null),
  feature_id: z.string().nullable().default(null),
  poi_id: z.string().nullable().default(null),
  trip_id: z.string().nullable().default(null),
  trip_title: z.string().nullable().default(null),
  external_id: z.string().nullable().default(null),
  address: z.string().nullable().default(null),
  road_address: z.string().nullable().default(null),
  category: z.string().nullable().default(null),
  marker_color: z.string().nullable().default(null),
  marker_icon: z.string().nullable().default(null),
  provider_url: z.string().nullable().default(null),
  phone: z.string().nullable().default(null),
});
export type PlaceSearchResult = z.infer<typeof PlaceSearchResultSchema>;

/** `GET /search` 응답 — internal → kakao → naver 순 정렬 + 소스별 degrade. */
export const PlaceSearchResponseSchema = z.object({
  results: z.array(PlaceSearchResultSchema),
  degraded_sources: z.array(z.string()),
});
export type PlaceSearchResponse = z.infer<typeof PlaceSearchResponseSchema>;
